crates/core/src/operations/connect.rs

```diff
@@ -864,56 +888,84 @@ pub(crate) async fn initial_join_procedure(
             gateways.len()
         );
 
+        let mut in_flight_gateways = HashSet::new();
+
         loop {
             let open_conns = op_manager.ring.open_connections();
+            let reserved_conns = op_manager
+                .ring
+                .connection_manager
+                .get_reserved_connections();
             let unconnected_gateways: Vec<_> =
                 op_manager.ring.is_not_connected(gateways.iter()).collect();
+            let available_gateways: Vec<_> = unconnected_gateways
+                .into_iter()
+                .filter(|gateway| !in_flight_gateways.contains(&gateway.peer))
+                .collect();
 
             tracing::debug!(
-                "Connection status: open_connections = {}, unconnected_gateways = {}",
-                open_conns,
-                unconnected_gateways.len()
+                open_connections = open_conns,
+                reserved_connections = reserved_conns,
+                inflight_gateway_dials = in_flight_gateways.len(),
+                available_gateways = available_gateways.len(),
+                "Connection status before join attempt"
             );
 
-            let unconnected_count = unconnected_gateways.len();
+            if reserved_conns > 0 {
```

Comment:
why are we doing this change? in theory we should be attempting to connect to multiple **different** gateways in parallel. what we should be trying is to not connect to the same gateway more than once